torch_agents/cogment_verse_torch_agents/muzero/adapter.py

```diff
@@ -0,0 +1,806 @@
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import asyncio
+from collections import defaultdict, namedtuple
+import ctypes
+import copy
+import time
+
+import logging
+import torch
+import torch.multiprocessing as mp
+from multiprocessing.pool import ThreadPool
+import numpy as np
+import queue
+
+from data_pb2 import (
+    MuZeroTrainingRunConfig,
+    MuZeroTrainingConfig,
+    AgentAction,
+    TrialConfig,
+    ActorParams,
+    EnvironmentConfig,
+    EnvironmentParams,
+    ActorConfig,
+)
+
+from cogment_verse.utils import LRU
+from cogment_verse import AgentAdapter
+from cogment_verse import MlflowExperimentTracker
+from cogment_verse_torch_agents.wrapper import np_array_from_proto_array, proto_array_from_np_array
+from cogment_verse_torch_agents.muzero.replay_buffer import Episode, TrialReplayBuffer, EpisodeBatch
+from cogment_verse_torch_agents.muzero.agent import MuZeroAgent
+
+from cogment.api.common_pb2 import TrialState
+import cogment
+
+
+# pylint: disable=arguments-differ
+
+log = logging.getLogger(__name__)
+
+
+MuZeroSample = namedtuple("MuZeroSample", ["state", "action", "reward", "next_state", "done", "policy", "value"])
+
+
+class LinearScheduleWithWarmup:
+    """Defines a linear schedule between two values over some number of steps.
+
+    If updated more than the defined number of steps, the schedule stays at the
+    end value.
+    """
+
+    def __init__(self, init_value, end_value, total_steps, warmup_steps):
+        """
+        Args:
+            init_value (Union[int, float]): starting value for schedule.
+            end_value (Union[int, float]): end value for schedule.
+            steps (int): Number of steps for schedule. Should be positive.
+        """
+        self._warmup_steps = max(warmup_steps, 0)
+        self._total_steps = max(total_steps, self._warmup_steps)
+        self._init_value = init_value
+        self._end_value = end_value
+        self._current_step = 0
+        self._value = 0
+
+    def get_value(self):
+        return self._value
+
+    def update(self):
+        # pylint: disable=invalid-name
+        if self._current_step < self._warmup_steps:
+            t = np.clip(self._current_step / (self._warmup_steps + 1), 0, 1)
+            self._value = self._init_value * t
+        else:
+            t = np.clip(
+                (self._current_step - self._warmup_steps) / max(1, self._total_steps - self._warmup_steps), 0, 1
+            )
+            self._value = self._init_value + t * (self._end_value - self._init_value)
+
+        self._current_step += 1
+        return self._value
+
+
+class RunningStats:
```

Comment:
Same should be extracted to its own file

Reply:
Done